### ISSUE TITLE
Backprop does not work for lfilter

### ISSUE THREAD

**Issue by kevinmchu** [pytorch/audio#1517]
## 🐛 Bug

Gradients are not being backpropagated when using lfilter. 
## To Reproduce

1. `filtered = F.lfilter(x, a, b)`

The command `list(model.parameters())[0].grad` returns `None` and the network weights are exactly the same after iterating through each batch. 

Output of `list(model.parameters())[0].grad`:
None
None
None
...

Model weights:
tensor([[ 0.0322,  0.0787, -0.0834,  ..., -0.0701, -0.0773, -0.0394],
        [-0.0378, -0.0445,  0.0642,  ...,  0.0215,  0.0875, -0.0312],
        [ 0.0328,  0.0955, -0.0069,  ...,  0.0831, -0.0243,  0.0969],
        ...,
        [ 0.0587, -0.0371,  0.0228,  ..., -0.0462,  0.0787, -0.0234],
        [ 0.0719,  0.0537,  0.0785,  ..., -0.0930, -0.0149, -0.0980],
        [-0.0099, -0.0187,  0.0200,  ..., -0.0295,  0.0229, -0.0406]],
       device='cuda:0')

tensor([[ 0.0322,  0.0787, -0.0834,  ..., -0.0701, -0.0773, -0.0394],
        [-0.0378, -0.0445,  0.0642,  ...,  0.0215,  0.0875, -0.0312],
        [ 0.0328,  0.0955, -0.0069,  ...,  0.0831, -0.0243,  0.0969],
        ...,
        [ 0.0587, -0.0371,  0.0228,  ..., -0.0462,  0.0787, -0.0234],
        [ 0.0719,  0.0537,  0.0785,  ..., -0.0930, -0.0149, -0.0980],
        [-0.0099, -0.0187,  0.0200,  ..., -0.0295,  0.0229, -0.0406]],
       device='cuda:0')

tensor([[ 0.0322,  0.0787, -0.0834,  ..., -0.0701, -0.0773, -0.0394],
        [-0.0378, -0.0445,  0.0642,  ...,  0.0215,  0.0875, -0.0312],
        [ 0.0328,  0.0955, -0.0069,  ...,  0.0831, -0.0243,  0.0969],
        ...,
        [ 0.0587, -0.0371,  0.0228,  ..., -0.0462,  0.0787, -0.0234],
        [ 0.0719,  0.0537,  0.0785,  ..., -0.0930, -0.0149, -0.0980],
        [-0.0099, -0.0187,  0.0200,  ..., -0.0295,  0.0229, -0.0406]],
       device='cuda:0')

...

## Expected behavior

The command `list(model.parameters())[0].grad` should return a matrix and the network weights should get updated for each iteration.

## Environment

 - What commands did you used to install torchaudio (conda/pip/build from source)?
`pip install torchaudio`

 - If you are building from source, which commit is it?
n/a

 - What does `torchaudio.__version__` print? (If applicable)
0.8.1

 - PyTorch Version (e.g., 1.0): 1.8.1
 - OS (e.g., Linux): Ubuntu 18.04
 - Python version: Python 3.6.12
 - CUDA/cuDNN version: 11.2
 - GPU models and configuration: NVIDIA TITAN V

**Comment by mthrok** [comment]
Hi @kevinmchu 

Can you provide a small snippet that shows how you are computing backward pass?
Specifically, did you register your coefficients `a` and `b` as buffer, and are you sure that `list(model.parameters())[0]` is the one you are passing to `lfilter`?

**Comment by mthrok** [comment]
This works;

```python
import torch
import torchaudio.functional as F


class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.register_buffer('a', torch.tensor([0.7, 0.2, 0.6]))
        self.register_buffer('b', torch.tensor([0.4, 0.2, 0.9]))

        self.a.requires_grad = True
        self.b.requires_grad = True

    def forward(self, x):
        return F.lfilter(x, self.a, self.b)



module = MyModule()

x = torch.randn(2, 255)
x.requires_grad = True

y = module(x)
loss = y.abs().square().sum()
loss.backward()

print(x.grad)
print(module.a.grad)
print(module.b.grad)
```

```
$ python foo.py
tensor([[ 1.4558e+00,  1.6507e+00, -6.1283e-01, -1.1323e+00,  1.4560e-01,
          8.7836e-01, -1.1573e+00, -1.2907e+00,  1.6356e+00, -1.4654e+00,
         -1.8944e+00,  1.1532e+00,  1.0003e+00, -1.9565e+00,  1.3916e+00,
          ...
          2.1292e+00,  1.2765e-01,  1.7035e-01,  3.9363e-01, -6.4180e-01,
         -5.8097e-01,  7.4801e-01,  5.3445e-01, -8.3886e-01,  1.7232e-01,
          1.2559e+00, -9.4408e-01,  5.0859e-01, -6.4047e-01,  0.0000e+00]])
tensor([-442.6186,  -53.7205,  295.7253])
tensor([-161.7525,   31.3564,  223.9684])
```

**Comment by kevinmchu** [comment]
I'm writing a custom loss function that includes use of `lfilter`. `x` is a 2D tensor, and `a` and `b` are 1D tensors. I'm computing the backward pass using `loss.backward()`.

**Comment by mthrok** [comment]
@kevinmchu Please look at my script and make sure that the `requires_grad` is set to `True` (for gradient computation) and the parameter is registered with `register_parameter`, (for recognized as parameter).

**Comment by mthrok** [comment]
@kevinmchu BTW autograd support to `lfilter` was added only recently. So you need nightly build of `PyTorch` and `torchaudio`. The PyTorch 1.8.1 you have does not work.

**Comment by kevinmchu** [comment]
I tried installing the nightly build, but I'm receiving messages that the requirements are already satisfied. Do you know what I'm doing wrong?

pip3 install --pre torch torchvision torchaudio -f https://download.pytorch.org/whl/nightly/cu112/torch_nightly.html
Looking in links: https://download.pytorch.org/whl/nightly/cu112/torch_nightly.html
Requirement already satisfied: torch in /xxx/anaconda3/lib/python3.6/site-packages (1.8.1)
Requirement already satisfied: torchvision in /xxx/anaconda3/lib/python3.6/site-packages (0.8.2)
Requirement already satisfied: torchaudio in /xxx/anaconda3/lib/python3.6/site-packages (0.8.1)
Requirement already satisfied: typing-extensions in /xxx/anaconda3/lib/python3.6/site-packages (from torch) (3.7.4.3)
Requirement already satisfied: dataclasses in /xxx/anaconda3/lib/python3.6/site-packages (from torch) (0.7)
Requirement already satisfied: numpy in /xxx/anaconda3/lib/python3.6/site-packages (from torch) (1.19.2)
Requirement already satisfied: pillow>=4.1.1 in /xxx/anaconda3/lib/python3.6/site-packages (from torchvision) (8.1.0)
WARNING: Error parsing requirements for astor: [Errno 2] No such file or directory: '/xxx/anaconda3/lib/python3.6/site-packages/astor-0.7.1.dist-info/METADATA'

**Comment by mthrok** [comment]
It's typical dependency issue. I would first uninstall the existing installation and try it again.

Also the log says 
```
Requirement already satisfied: torch in /xxx/anaconda3/lib/python3.6/site-packages (1.8.1)
Requirement already satisfied: torchvision in /xxx/anaconda3/lib/python3.6/site-packages (0.8.2)
Requirement already satisfied: torchaudio in /xxx/anaconda3/lib/python3.6/site-packages (0.8.1)
```

So you do need to uninstall all `torch=1.8.1`, `torchaudio=0.8.1` and `torchvision=0.8.2`.

**Comment by kevinmchu** [comment]
I installed the nightly build via
```
pip uninstall torch torchvision torchaudio
pip3 install --pre torch torchvision torchaudio -f https://download.pytorch.org/whl/nightly/cu112/torch_nightly.html
```

but still had issues with backprop.

**Comment by mthrok** [comment]
Can you show the relevant part of the code?

**Comment by kevinmchu** [comment]
Simplified version of class definition:

```
class Filter(torch.nn.Module):
    def __init__(self):
        super().__init__()

        ###
        Code to calculate a and b, which are (8x3) tensors 
        ###

	self.register_parameter('a', torch.nn.Parameter(a, requires_grad=True))
        self.register_parameter('b', torch.nn.Parameter(b, requires_grad=True))

    def forward(self, x, m):
        # x is a 2D tensor, m is an index

        return F.lfilter(x, self.a[m, :], self.b[m, :])
```

**Comment by mthrok** [comment]
To compute gradient, you need to pass the whole Tensor. You cannot compute the gradient for a part of Tensor.

You can work around the by treating each filter separately.

```python
import torch
import torchaudio.functional as F


class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()

        self.a = torch.nn.ParameterList()
        self.b = torch.nn.ParameterList()
        for i in range(8):
            v = i + 1.
            tensor = torch.tensor([v, v, v], dtype=torch.float32)
            self.a.append(torch.nn.Parameter(tensor, requires_grad=True))
            tensor = -torch.tensor([v, v, v], dtype=torch.float32)
            self.b.append(torch.nn.Parameter(tensor, requires_grad=True))

    def forward(self, x, i):
        return F.lfilter(x, self.a[i], self.b[i])


module = MyModule()

x = torch.ones(2, 5)

for i in range(8):
    print('Index:', i)
    y = module(x, i)
    loss = y.abs().square().sum()
    loss.backward()

    for name, param in module.named_parameters():
        print(f' - {name}:', param.grad)
```

<details>

```
$python foo.py
Index: 0
 - a.0: tensor([-8., -8., -4.])
 - a.1: None
 - a.2: None
 - a.3: None
 - a.4: None
 - a.5: None
 - a.6: None
 - a.7: None
 - b.0: tensor([-8., -8., -4.])
 - b.1: None
 - b.2: None
 - b.3: None
 - b.4: None
 - b.5: None
 - b.6: None
 - b.7: None
Index: 1
 - a.0: tensor([-8., -8., -4.])
 - a.1: tensor([-4., -4., -2.])
 - a.2: None
 - a.3: None
 - a.4: None
 - a.5: None
 - a.6: None
 - a.7: None
 - b.0: tensor([-8., -8., -4.])
 - b.1: tensor([-4., -4., -2.])
 - b.2: None
 - b.3: None
 - b.4: None
 - b.5: None
 - b.6: None
 - b.7: None
Index: 2
 - a.0: tensor([-8., -8., -4.])
 - a.1: tensor([-4., -4., -2.])
 - a.2: tensor([-2.6667, -2.6667, -1.3333])
 - a.3: None
 - a.4: None
 - a.5: None
 - a.6: None
 - a.7: None
 - b.0: tensor([-8., -8., -4.])
 - b.1: tensor([-4., -4., -2.])
 - b.2: tensor([-2.6667, -2.6667, -1.3333])
 - b.3: None
 - b.4: None
 - b.5: None
 - b.6: None
 - b.7: None
Index: 3
 - a.0: tensor([-8., -8., -4.])
 - a.1: tensor([-4., -4., -2.])
 - a.2: tensor([-2.6667, -2.6667, -1.3333])
 - a.3: tensor([-2., -2., -1.])
 - a.4: None
 - a.5: None
 - a.6: None
 - a.7: None
 - b.0: tensor([-8., -8., -4.])
 - b.1: tensor([-4., -4., -2.])
 - b.2: tensor([-2.6667, -2.6667, -1.3333])
 - b.3: tensor([-2., -2., -1.])
 - b.4: None
 - b.5: None
 - b.6: None
 - b.7: None
Index: 4
 - a.0: tensor([-8., -8., -4.])
 - a.1: tensor([-4., -4., -2.])
 - a.2: tensor([-2.6667, -2.6667, -1.3333])
 - a.3: tensor([-2., -2., -1.])
 - a.4: tensor([-1.6000, -1.6000, -0.8000])
 - a.5: None
 - a.6: None
 - a.7: None
 - b.0: tensor([-8., -8., -4.])
 - b.1: tensor([-4., -4., -2.])
 - b.2: tensor([-2.6667, -2.6667, -1.3333])
 - b.3: tensor([-2., -2., -1.])
 - b.4: tensor([-1.6000, -1.6000, -0.8000])
 - b.5: None
 - b.6: None
 - b.7: None
Index: 5
 - a.0: tensor([-8., -8., -4.])
 - a.1: tensor([-4., -4., -2.])
 - a.2: tensor([-2.6667, -2.6667, -1.3333])
 - a.3: tensor([-2., -2., -1.])
 - a.4: tensor([-1.6000, -1.6000, -0.8000])
 - a.5: tensor([-1.3333, -1.3333, -0.6667])
 - a.6: None
 - a.7: None
 - b.0: tensor([-8., -8., -4.])
 - b.1: tensor([-4., -4., -2.])
 - b.2: tensor([-2.6667, -2.6667, -1.3333])
 - b.3: tensor([-2., -2., -1.])
 - b.4: tensor([-1.6000, -1.6000, -0.8000])
 - b.5: tensor([-1.3333, -1.3333, -0.6667])
 - b.6: None
 - b.7: None
Index: 6
 - a.0: tensor([-8., -8., -4.])
 - a.1: tensor([-4., -4., -2.])
 - a.2: tensor([-2.6667, -2.6667, -1.3333])
 - a.3: tensor([-2., -2., -1.])
 - a.4: tensor([-1.6000, -1.6000, -0.8000])
 - a.5: tensor([-1.3333, -1.3333, -0.6667])
 - a.6: tensor([-1.1429, -1.1429, -0.5714])
 - a.7: None
 - b.0: tensor([-8., -8., -4.])
 - b.1: tensor([-4., -4., -2.])
 - b.2: tensor([-2.6667, -2.6667, -1.3333])
 - b.3: tensor([-2., -2., -1.])
 - b.4: tensor([-1.6000, -1.6000, -0.8000])
 - b.5: tensor([-1.3333, -1.3333, -0.6667])
 - b.6: tensor([-1.1429, -1.1429, -0.5714])
 - b.7: None
Index: 7
 - a.0: tensor([-8., -8., -4.])
 - a.1: tensor([-4., -4., -2.])
 - a.2: tensor([-2.6667, -2.6667, -1.3333])
 - a.3: tensor([-2., -2., -1.])
 - a.4: tensor([-1.6000, -1.6000, -0.8000])
 - a.5: tensor([-1.3333, -1.3333, -0.6667])
 - a.6: tensor([-1.1429, -1.1429, -0.5714])
 - a.7: tensor([-1.0000, -1.0000, -0.5000])
 - b.0: tensor([-8., -8., -4.])
 - b.1: tensor([-4., -4., -2.])
 - b.2: tensor([-2.6667, -2.6667, -1.3333])
 - b.3: tensor([-2., -2., -1.])
 - b.4: tensor([-1.6000, -1.6000, -0.8000])
 - b.5: tensor([-1.3333, -1.3333, -0.6667])
 - b.6: tensor([-1.1429, -1.1429, -0.5714])
 - b.7: tensor([-1.0000, -1.0000, -0.5000])
```

</details>

If you need to recover the filter in one Tensor, you can do something like `torch.cat([t for t in module.a])`. This can construct a Tensor with the corresponding filter coefficients, but the corresponding gradients are not retained, and I do not know a way to construct a Tensor while holding their gradients. I am not sure if that's what PyTorch is designed for.

**Comment by kevinmchu** [comment]
Is there a function in torchaudio or pytorch that would allow me to apply a bank of 8 iir filters all at once without having to write a for loop?

**Comment by mthrok** [comment]
> Is there a function in torchaudio or pytorch that would allow me to apply a bank of 8 iir filters all at once without having to write a for loop?

I am not aware of one.  @yoyololicon Do you know if there is any?

Just to be clear, can you give an example of shape?
I think your input is `[batch, frame]` and filter coefficients are `[num_filter, filter_dim]`, what's your desired output shape?

**Comment by kevinmchu** [comment]
My input is `[x, num_frames]` where x is a hyperparameter and the desired output shape is `[x, num_filters, num_frames]`. I wrote a for loop to calculate the loss of each sample in the batch separately, though I just realized this is another reason autograd is not working.

**Comment by mthrok** [comment]
> My input is `[x, num_frames]` where x is a hyperparameter and the desired output shape is `[x, num_filters, num_frames]`. I wrote a for loop to calculate the loss of each sample in the batch separately, though I just realized this is another reason autograd is not working.

I do not think it's necessary to loop through samples in batch. It should be possible to pass the batch, though you still need to loop through filters registered as `ParameterList`. Despite that, I think it's a good addition to torchaudio to support general filter bank application.

**Comment by yoyololicon** [comment]
> > Is there a function in torchaudio or pytorch that would allow me to apply a bank of 8 iir filters all at once without having to write a for loop?
> 
> I am not aware of one. @yoyololicon Do you know if there is any workaround?

Unfortunately, for-loop is probably the only solution for now.
We do plan to support batched filter coefficients (see #1476 ), but it's a little bit different from your application.
I agree that multiple filter banks would be a good add-on feature.
